File: coloring-enumeration/enumerate.cpp

#include <iostream>
#include <set>
#include <vector>

using namespace std;

typedef std::set<string> Set;
typedef std::vector<int> Vector;


string encode(Vector partial) {
	string b = "";
	for (int k : partial) b = b + std::to_string(k);

	return b;
}


Set colorings(Vector partial, Vector counts, Vector Q, int q) {
	Set include = Set();

	// Check whether we have equality in the number of colors. If so, return
	// now.
	bool equality = true;
	for (int i=0; i<q; i++) equality &= counts[i] == Q[i];

	if (equality) {
		include.insert(encode(partial));
		return include;
	}

	// Otherwise, build new ints and keep trying.
	for (int digit=0; digit<q; digit++) {
		Vector proposed(counts);
		proposed[digit] += 1;

		// Check whether we're staying in-bounds...
		bool inbounds = true;
		for (int i=0; i<q; i++) inbounds &= Q[i] - proposed[i] >= 0;

		// ... if we aren't, continue to the next proposed digit...
		if (!inbounds) continue;

		// ... and if we are, recurse.
		Vector partialplus(partial);
		partialplus.push_back(digit);

		Set more = colorings(partialplus, proposed, Q, q);
		std::set_union(more.begin(), more.end(), include.begin(), include.end(), std::inserter(include, include.end()));
	}

	return include;
}

int main(int argc, char *argv[]) {
	std::vector<string> args(argv+1, argv+argc);

	Vector Q = Vector();
	for (auto str : args) Q.push_back(std::stoi(str));
	
	Vector partial = Vector();
	int q = Q.size();
	Vector counts = Vector(q, 0);

	Set colors = colorings(partial, counts, Q, q);

	for (auto it : colors) {
		std::cout << it << std::endl;
	}

	std::cout << colors.size() << std::endl;

	return 0;
}

File: coloring-enumeration/enumerate.py

import numpy as np
from scipy.special import factorial

def multi(Q):
	Q = np.array(Q)
	return int(factorial(Q.sum())/factorial(Q).prod())

def baseq(part, q): return sum(part[k]*(q**k) for k in range(len(part)))


def digits(part, counts, Q):
	include = set()

	# If the counts are correct, return the partial (now complete) string.
	if counts == Q:
		return set([baseq(part, 10)])

	# Otherwise, recurse.
	for digit in range(len(Q)):
		# Propose the partial demography; if it violates the global one, continue.
		proposed = tuple(counts[i] if i != digit else counts[i] + 1 for i in range(len(counts)))
		if any(Q[i]-counts[i] < 0 for i in range(len(Q))): continue

		# Otherwise, recurse and add more digits.
		include |= digits(part + [digit], proposed, Q)
	
	return include


# Number of colors; demography; number of vertices.
q = 3
Q = (1, 4, 5)
N = 10

# Assert that the demography lines up with the number of colors.
assert sum(Q) == N

# Compute the different colorings and assert we have the right number of them.
S = digits([], (0,0,0), Q)
should = multi(Q)

assert len(S) == should
